### DISCUSSION TITLE
Support for non-Java JVM scripting languages in Ghidra Scripts

### DISCUSSION THREAD

**Discussion by broosa** [NationalSecurityAgency/ghidra#2886] [Ideas]
Hey all!

Going to tag @jpleasu here as he seems to have some familiarity with this portion of the Ghidra codebase...

As mentioned in #2490 it looks like Ghidra 9.2 has moved to a different method for run-time compilation of scripts. I'd love to see support for non-Java languages (a la Groovy, Scala, etc). Is this something thats being worked on now? I may be able to provide some cycles to contribute to the effort, but don't want to conflict with any other efforts if they're already underway.

Looking at the code for `GhidraSourceBundle` and the other OSGi functionality, it seems that it may be possible with a bit of work to add a step for compiling .groovy files just like .java files are compiled. Does anyone have any insight into what else might be required   to add such functionality?

**Comment by jpleasu** [comment]
I don't know of any other interest in OSGi aware script providers for non-Java languages.

There are a few parts necessary to participate in Ghidra's OSGi aware scripting:
1. **change detection**. when source files are newer than their corresponding classes (inner classes too) they need to be recompiled.  This might be more complicated for non-Java languages where support classes are generated.
<br/> I still don't correctly handle the case where multiple top-level classes are defined in a single .java file.  Getting it right would mean parsing the source and the balance between accuracy, complexity, and speed doesn't favor doing that.
2. **an OSGI aware compiler**.  `GhidraSourceBundle` uses [Phidias](https://github.com/rotty3000/phidias).  An approximation would be to feed your language's compiler a classpath generated from the currently active bundles.
3. **a manifest**.  `GhidraSourceBundle` will create a manifest (if there's isn't one in the source directory)  from the results of compilation using [bndtools](https://github.com/bndtools/bnd).  The language runtime packages would end up as imports here.
4. **language runtime**.  E.g.  `scala-library.jar` or `groovy.jar`.  These jars already include OSGi metadata, so they could be loaded as bundles.  They could also be included in a module's `lib` directory and found by `ClassSearcher` then provided through the system bundle in the same way that "legacy" Ghidra packages are now.

Factoring out the current Java implementation to a new `JVMLanguageProvider` extension point would be done in Ghidra, the non-java languages would be (external) extensions.

**Reply by broosa** [reply]
> This might be more complicated for non-Java languages where support classes are generated.

Do you have a specific example of this? I'm honestly kinda curious; hadn't taken this part into consideration.



> an OSGI aware compiler. GhidraSourceBundle uses Phidias. 

Oh- this reminds me of a question I had: am I correct in thinking that Phidias resolve inter-bundle dependencies, while intra-bundle dependencies are resolved as by the compiler as part of the actual compiler execution (through use of the source path and class path options)?

> An approximation would be to feed your language's compiler a classpath generated from the currently active bundles.

Yeah, that's kinda what I was thinking of as well. I think in Groovy's case the compilation steps mimic Java very closely, so this would be the best bet.

> Factoring out the current Java implementation to a new JVMLanguageProvider extension point would be done in Ghidra, the non-java languages would be (external) extensions.

I was thinking of something along these very lines... the language provider would mainly need to be responsible for generating .class files from corresponding language source code. I could even see the possibility of mixing multiple languages in the same source directory, then picking the right compiler based on the file extension. I think the biggest hurdle would be intra-bundle dependency handling.

I can see a given language provider implementing the following methods, on top of what's already in the source code:

- A method that, given a source code file, generates a list of classes that the file depends on (its "requirements") based on its import statements
- A method that, given a source code file, generates a list of classes defined in that source code file (its "capabilities")
- A method that, given a source code file, compiles it to a .class file 

The bundle compilation process would then go something like this:

1. For each file, select an appropriate language provider based on metadata, file extension, etc, and generate a list of capabilities and requirements for that file. Eventually you'd end up with a mapping between classes and files as well as a mapping between each file and the others it depended on.
2. Generate a dependency tree and determine a compilation order that ensures all dependencies are satisfied.
3. In the order generated in step 2, compile each file with the language-specific compiler process, with the output being placed in a common classpath so that they can be used in other files and/or languages.

I don't think this would require any changes to inter-bundle dependency management, as I think that would be handled already once the code is compiled.

Anyhow, I'd love to hear your thoughts on that approach.

**Reply by broosa** [reply]
Just to add a bit regarding compilation between languages- it looks like Gradle's Groovy and Scala plugins both support joint compilation, allowing mixing of Gradle/Scala code with java. Is this something that could be leveraged to remove some of the effort needed to handle compilation and file dependencies?

**Reply by jpleasu** [reply]
> Do you have a specific example of this? I'm honestly kinda curious; hadn't taken this part into consideration.
> 
`1.scala`:
```scala
package q {
  object a {
  }

  class X() {
  }
}

package r {
  object b {
  }

  class Y() {
  }

  package s {
    object c {
    }

    class Z() {
    }
  }
}
```
take a look at the class files generated by `scalac 1.scala`.

> Oh- this reminds me of a question I had: am I correct in thinking that Phidias resolve inter-bundle dependencies, while intra-bundle dependencies are resolved as by the compiler as part of the actual compiler execution (through use of the source path and class path options)?

almost.. during change detection, "old" class files are deleted.  All of the source without current class files are handed to the compiler, and any remaining class files are on the classpath provided to the compiler.

> 
> I can see a given language provider implementing the following methods, on top of what's already in the source code:
> 
>     * A method that, given a source code file, generates a list of classes that the file depends on (its "requirements") based on its import statements
> 
What about:
```java
import java.util.*;
```
or the use of classes at package scope?  they don't have corresponding imports.

You need to actually parse the file.  At least Java doesn't have dependent names!


>     * A method that, given a source code file, generates a list of classes defined in that source code file (its "capabilities")
> 
This is the parsing I referred to above that I'm resistant to.  I started down this path as a way to solve the multiple top-level class problem for Java, but decided against it.  Introducing code to parse and analyze script source increases maintenance burden quite a bit and the benefit (in my case) would be very small.  I think it's a hard sell.


>     * A method that, given a source code file, compiles it to a .class file
> 
One source file can produce a _bunch_ of .class files (lambdas and other anonymous classes, nested classes, and non-java "support" clases).  The current method "handles" intra-bundle dependency cycles by compiling the set of new source files at once.

>     2. Generate a dependency tree and determine a compilation order that ensures all dependencies are satisfied.
> 
do you forbid cycles?


> I don't think this would require any changes to inter-bundle dependency management, as I think that would be handled already once the code is compiled.
> 
It's handled by creating the manifest that tells OSGi how to handle it.  That's currently done with bndtool and bndtool takes configuration.  E.g. see the method `GhidraSourceBundle#generateManifest`.  I've exported all but "internal" and "private" packages (and their subpackages) -- I'm not sure if that's appropriate for other JVM based languages.  A language provider would need some form of participation in the creation of the manifest.

**Comment by fmagin** [comment]
@broosa do you care about the whole fancy OSGi aware scripting, or do you just want scripts in other JVM languages?
Because there are already extensions for a [Scala Script Provider](https://github.com/edmcman/ghidra-scala-loader) and a [Kotlin Script Provider](https://github.com/GhidraJupyter/ghidra-jupyter-kotlin). The Scala Script Provider just calls the Scala Compiler via a subrprocess, the Kotlin Provider used to do that, but nowadays uses the the actual runtime API that the Kotlin Ecosystem provides for that. I think both are not as fully thought out as the new JavaScriptProvider (like detecting changed files), but they should work for many basic use cases.

(Disclaimer: I am one of the authors of the Kotlin Support Extension for Ghidra)

**Reply by goatshriek** [reply]
Another option is the [RubyDragon](https://github.com/goatshriek/ruby-dragon) extension I maintain which supports Ruby, Clojure, and Kotlin scripts. It also provides REPL windows for each of these, similar to what is built-in for Python.
